RI Claim Embedded Tokens

cdvoisin opened the pull request:
RI Claim Embedded Passport Tokens removed.

cdvoisin committed:
RI Claim Embedded Tokens

RI Claim Embedded Passport Tokens removed.

cdvoisin committed:
Add ga4gh.LinkedIdentities claim

cdvoisin committed:
Add LinkedIdentities to full example

Note that the other claims in the example didn't change except for adding "iss" and "sub" to AcceptedTermsAndPolicies as well as to ResearcherStatus. Most of the diff in the section is an indentation change.

mikael-linden reviewed (commented):
- definition of claim embedded token _“An Assertion from a Claim Authority that is bound to a researcher identity and signed by an Identity Broker as per the "iss" field.”_ should it be “signed by the Claim Authority” as I think that is the point here

- passport access token. Do we still believe in the approach where access tokens can carry claims as the size of embedded token is quite big 

- Exp definition: _“REQUIRED. Generally, it is seconds since unix epoch of when the Claim Authority requires such a claim to be no longer valid. A Claim Authority MAY choose to make a claim very long lived. However, an Identity Broker MAY choose an earlier timestamp if it wishes to limit the claim’s duration of use within downstream Claim Clearinghouses.”_ How can a broker change the timestamp of a claim that is made by a claim authority?

cdvoisin committed:
Remove claims on access token

And rename Claim Authority to be Claim Source.

cdvoisin commented:
Mikael, I worked with David Bernick on Friday to add new definitions and service roles to the AAI spec to give us a more well-rounded picture of where claims come from: https://github.com/ga4gh/data-security/blob/master/AAI/AAIConnectProfile.md#flow-of-claims. Also defined these newly detailed services: https://github.com/ga4gh/data-security/blob/master/AAI/AAIConnectProfile.md#terminology.

With this new context and with an additional RI edit included in this PR: a Claim Source is an org that mints claims before Brokers do their primary work (although it is possible a Broker is a mega service that does all of these things). Claims are stored in a Claim Repository such as a database. An Embedded Claim Signatory service reads the claims from the database and signs them. This may or may not be the Broker itself. Finally, a Broker bundles together these signed claims and signs a Passport Access Token (that contains no RI Claims) and makes the RI Claims available only via the /userinfo endpoint.

As for the three comments from Mikael:

1. I'm trying to give context in the spec that the Claim Source is where claims come from, and ultimately the responsible party for why the claim exists in the first place. The previous term "Claim Authority" was too easy to mix up with the service that signs stuff, so that term has been replaced.

2. This CL and one submitted in the AAI spec now remove RI Claims from the access token. We have been talking about it for a while, now it is done.

3. "exp" used to have a Broker be able to reduce the time of expiry. Now with embedded RI claim objects, it is the Embedded Claim Signatory's option to do this. Obligation even if the Claim Signatory only uses "jku" and the expiry is more than 30 days in the future and it is the Claim Signatory's job to revoke Embedded Tokens if it issues Embedded Tokens with the "openid" scope such that it is allowed to issue an "exp" more than 30 days. Doing this makes the Claim Signatory more and more like a Broker and dealing with the world of AAI tokens, which is fine. Issuing "jku" embedded claims is the other way to go and does not require all of that extra AAI inside the Claim Signatory service.

With this new context and PRs up to date, PTAL at this PR. Thank you!

martin-kuba commented on researcher_ids/RI_Claims_V1.md:
```diff
@@ -60,90 +65,85 @@ The key words "MUST", "MUST NOT", "REQUIRED", "SHALL", "SHALL NOT", "SHOULD",
+    
+-   RI Claim Embedded Tokens are encoded as strings and included as a list of
+    strings returned within a [RI Claim](#researcher-identity-claim-ri-claim). See
+    the [RI Claim Embedded Token](#ri-claim-embedded-token) section of this
```
The name of the current section _Researcher Identity Claim Embedded Token ("RI Claim Embedded Token")_ is the same as the name of the referenced section _RI Claim Embedded Token_  which is a bit confusing.

cdvoisin replied:
Renamed the section that describes the format to use "format" in its title.

cdvoisin replied:
Renamed sections.

martin-kuba commented on researcher_ids/RI_Claims_V1.md:
```diff
@@ -60,90 +65,85 @@ The key words "MUST", "MUST NOT", "REQUIRED", "SHALL", "SHALL NOT", "SHOULD",
--   The
-    [AAI Claim Authority](https://github.com/ga4gh/data-security/blob/master/AAI/AAIConnectProfile.md#term-claim-authority)
-    as encoded by the [source field](#source-required).
+An OIDC claim "ga4gh_rio" that contains fields that provide GA4GH
```
The ga4gh_rio claim **is not**  an OIDC claim. It is a JWT claim as specified in the section 2 "Terminology" of RFC 7519 "JSON Web Token", see https://tools.ietf.org/html/rfc7519#section-2
OIDC claims as defined in https://openid.net/specs/openid-connect-core-1_0.html#Claims are defined as members of JSON object returned by /userinfo or in id_token.

Thus this section should be reworded so that "OIDC" should be replaced with "JWT", and it should mention the _JSON Web Token Claims Registry_ 
https://tools.ietf.org/html/rfc7519#section-10.1 instead of OIDC specification.

cdvoisin replied:
Done

martin-kuba commented on researcher_ids/RI_Claims_V1.md:
```diff
@@ -60,90 +65,85 @@ The key words "MUST", "MUST NOT", "REQUIRED", "SHALL", "SHALL NOT", "SHOULD",
-    well as any RI Claims that may be collected either directly on JWTs provided
-    within the "ga4gh_passports" object or within responses to the /userinfo
-    endpoints for the "ga4gh_passports" tokens.
+-   The {"iss", "sub"} pair of OIDC standard claims that are included
```
Again, "iss" and "sub" are not OIDC standard claims, they are JWT standard claims defined in the JWT specification https://tools.ietf.org/html/rfc7519#section-4.1.1

cdvoisin replied:
Done

martin-kuba commented on researcher_ids/RI_Claims_V1.md:
```diff
@@ -60,90 +65,85 @@ The key words "MUST", "MUST NOT", "REQUIRED", "SHALL", "SHALL NOT", "SHOULD",
-    [Embedded Token](https://github.com/ga4gh/data-security/blob/master/AAI/AAIConnectProfile.md#term-embedded-token).
+-   A logical collection of RI Claims that includes claims on the
+    [Passport Access Token](#passport-access-token) as well as any
+    indirect RI Claims that such a JWT token makes available to the
```
A **JWT token** does not make anything available to the caller of /userinfo endpoint. Only an **access token** can do that. Thus the word "JWT" should be replaced with the word "access".

cdvoisin replied:
Done

martin-kuba commented on researcher_ids/RI_Claims_V1.md:
```diff
@@ -60,90 +65,85 @@ The key words "MUST", "MUST NOT", "REQUIRED", "SHALL", "SHALL NOT", "SHOULD",
+    [GA4GH AAI specification](https://github.com/ga4gh/data-security/blob/master/AAI/AAIConnectProfile.md)
+    for the purpose of encoding identity and evaluating authorization.
+    
+-   A Passport Access Token does not contain any
```
This is contrary to the text of the previous section titled "Passport".

cdvoisin replied:
Good catch, thanks. Fixed.

martin-kuba commented on researcher_ids/RI_Claims_V1.md:
```diff
@@ -202,58 +205,62 @@ Additonal terms on fields:
+        based on the "scope" claim as outlined elsewhere within this
+        specification.
+
+    3.  The Identity Broker introduces new RI Claim Embedded Tokens from a
```
This sentence is confusing. A broker does not introduce embedded tokens by encoding claim objects, it simply includes embedded tokens received from Embedded Token Signatory without any modification into the RI Claims.

cdvoisin replied:
Improved the description.

martin-kuba commented on researcher_ids/RI_Claims_V1.md:
```diff
@@ -202,58 +205,62 @@ Additonal terms on fields:
+        [AAI Specification](https://github.com/ga4gh/data-security/blob/master/AAI/AAIConnectProfile.md).
+
+    4.  The Identity Broker propagates RI Claims that it received
+        from upstream Brokers by copying each RI Claim Object Embedded
```
```suggestion
        from upstream Brokers by copying each RI Claim Embedded
```

cdvoisin replied:
Done

martin-kuba commented on researcher_ids/RI_Claims_V1.md:
```diff
@@ -202,58 +205,62 @@ Additonal terms on fields:
+        OIDC claim. These strings are provided unmodified, however they
+        MAY be filtered as described elsewhere in this specification.
+
+    5.  The Identity Broker signs the JWT, making it a Passport Access
```
The section v. does not make sense at all. The broker returns RI Claims in the /userinfo response, which is not a JWT, and it cannot possibly be an access token.

cdvoisin replied:
This is the signing of the Passport Access Token, which is the Broker's responsibility. Reordered and rephrased some of this section to simplify to some degree.

martin-kuba commented on researcher_ids/RI_Claims_V1.md:
```diff
@@ -202,58 +205,62 @@ Additonal terms on fields:
+    5.  The Identity Broker signs the JWT, making it a Passport Access
+        Token.
+
+    6.  The Passport for the /userinfo request MUST contain "ga4gh" as a
```
The "Passport" as defined above is not an access token, thus "the passport for the /userinfo request" is a non-defined thing.
It should probably be "access token", but in fact there is no need to have the "scope" claim inside of the access token, it is not used by the Authorization Server itself.

cdvoisin replied:
Rephrased

martin-kuba commented on researcher_ids/RI_Claims_V1.md:
```diff
@@ -202,58 +205,62 @@ Additonal terms on fields:
+    2.  Claim Clearinghouses SHOULD ignore claims that aren’t needed for their
         purposes.
+        
+    3.  When combining RI Claim Objects from multiple RI Claim Accounts for
```
```suggestion
    3.  When combining RI Claim Embedded Tokens from multiple RI Claim Sources for
```

cdvoisin replied:
Rephrased

martin-kuba commented on researcher_ids/RI_Claims_V1.md:
```diff
@@ -279,19 +286,128 @@ interfaces, such as:
-#### "**value**" [required]
+-   RI Claim Embedded Tokens MUST conform to the
+    [GA4GH AAI Specification Embedded Token format](https://github.com/ga4gh/data-security/blob/master/AAI/AAIConnectProfile.md#embedded-token-issued-by-broker)
+    as a base64 string.
```
Is is not a base64 string, it is a "JWS Compact Serialization" string as defined at https://tools.ietf.org/html/rfc7515#section-7.1 i.e. it is 3 base64 strings joined by dots.

cdvoisin replied:
Done

martin-kuba commented on researcher_ids/RI_Claims_V1.md:
```diff
@@ -279,19 +286,128 @@ interfaces, such as:
+    Embedded Token encoding options, then the /userinfo endpoint is unused
+    but the /introspect endpoint MUST be available for
+    [claim polling](https://github.com/ga4gh/data-security/blob/master/AAI/AAIConnectProfile.md#claim-polling).
+
```
It is not a requirement for an RI Claim Embedded Token to contain an "RI Claim Object" in the ga4gh_rio JWT claim?

cdvoisin replied:
Removed this stuff as it is now covered better by the AAI spec.

martin-kuba commented on researcher_ids/RI_Claims_V1.md:
```diff
@@ -279,19 +286,128 @@ interfaces, such as:
+the outer JWT itself.
+
+One of `scope` or `jku` MUST be present as described in
+[RI Claim Embedded Token Requirements](#ri-claim-embedded-token-requirements).
```
The referenced section does not contain the mentioned requirement.

cdvoisin replied:
Fixed.

mikael-linden commented:
For clarity, I would separate claim presentation (syntax and semantics) and claim transport (how they are fetched, e.g. /userinfo or else) and make this spec focus on presentation 

- Can the “passport access token” concept be removed from this document
- Can references to /userinfo be removed from the spec

Reshuffling the sections:
-	would clear separation of actors (claim source, embedded claim signatory, OIDC broker, claims clearinghouse) and objects (passport, RI claim, embedded token, RI claim object) make it easier to understand? E.g. introduce the actors and objects in two separate sections
-	would top-down (passport, RI claim, embedded token, RI claim object) or bottom-up ordering of the object definitions serve easier understand?
-	“claim and token revocation” section might better fit next to “claim expiry” N.B. I understand “revocation” means “take back the claim before it expires” so expiry and revocation are different things.

lpaglione commented on researcher_ids/RI_Claims_V1.md:
```diff
@@ -60,90 +65,85 @@ The key words "MUST", "MUST NOT", "REQUIRED", "SHALL", "SHALL NOT", "SHOULD",
+    [AAI Claim Source](https://github.com/ga4gh/data-security/blob/master/AAI/AAIConnectProfile.md#term-claim-source)
+    as encoded by the [source field](#source).
+
+#### **Passport**
```
I know that a "Passport" is a somewhat conceptual idea, though its presence in the specifications always feels like it should be a specific data object with a specific data structure. I think from a data structure perspective we're referring to a bundle of claims as a token? Should we reconcile that language somewhat?

cdvoisin replied:
Although probably not what you were expecting, there has been a large overhaul of definitions including this one, as well as an Overview section to try to tie it together better.

cdvoisin committed:
Change "conditions" to "policies" 

In AcceptedTermsAndPolicies description

cdvoisin committed:
Reintroduce "asserted" RI Claim Object field

cdvoisin committed:
Addressing comments and formatting issues

cdvoisin committed:
Remove Passport Access Token

cdvoisin committed:
Redo definitions

cdvoisin committed:
Rename Access Assertion to Passport Visa

cdvoisin committed:
Rename JWT claim names for passport and visa

Also update examples to make "type" part of the Passport Visa Object.

cdvoisin committed:
Rename URL Claim Field to URL Field

cdvoisin committed:
Introduce Passport Clearinghouse

Make the separation more clear on which type of Claim Clearinghouse is conforming. It is one that handles Passports.

cdvoisin committed:
Introduce Repository, Signatory, Broker actors

For passports and visas.

cdvoisin committed:
Split definitions by actor and object

And rename "JWT Passport Claim" to "Passport JWT Claim".

cdvoisin committed:
Clarify services and orgs of definitions

cdvoisin committed:
Tweak URL Fields wording

cdvoisin committed:
Add basic passport flow of claims diagram

cdvoisin committed:
Delete passport_flow_of_claims_basic.svg

cdvoisin committed:
Add overview diagrams

cdvoisin committed:
Add signature to passport composition diagram

cdvoisin committed:
Introduce the Overview section

cdvoisin committed:
Passport Visa Source Repository description

Clarify "Passport Visa Source" Repository vs. "Passport Visa" Source Repository (i.e. confusion that it may be the source of visas instead of the source's repository.

cdvoisin committed:
Diagram rename to Passport Assertion Source

cdvoisin committed:
Rename to Passport Assertion Source

Also Passport Assertion Repository.

cdvoisin committed:
Fixed typos

Updated description of v0.9.5

cdvoisin committed:
More details on payload in examples

cdvoisin committed:
Update the README for new definitions

martin-kuba commented on researcher_ids/README.md:
```diff
@@ -140,14 +144,14 @@ Here is an example of the life of a claim:
 ## Why so many expiry timestamps?
 
-OIDC tokens have an "exp" standard claim to represent when the token expires. This is separate from the expiry time of claims themselves, which have  an "expires" field for when that claim expires. These different types of expiry timestamps serve different purposes:
+OIDC tokens have an "exp" standard claim to represent when the token expires. The Passport in an OIDC token and has an "exp" claim to indicate when it expires. This is separate from the expiry time of Passport Visas themselves, which each have an "exp" claim for when that Passport Visa expires. These different types of expiry timestamps serve different purposes:
```
There is no such thing as an "OIDC token". The OIDC spec inherits "access token" from the OAuth 2 spec, and adds id_token.

cdvoisin replied:
done

martin-kuba commented on researcher_ids/RI_Claims_V1.md:
```diff
@@ -60,200 +76,281 @@ The key words "MUST", "MUST NOT", "REQUIRED", "SHALL", "SHALL NOT", "SHOULD",
+-   A [GA4GH
+    Claim](https://github.com/ga4gh/data-security/blob/master/AAI/AAIConnectProfile.md#term-ga4gh-claim)
+    that is encoded as a list of [Passport Visas](#passport-visa)
+    provided by a common key value within the `ga4gh_passport_v1` JWT claim.
```
This line says "provided by a common key value within the `ga4gh_passport_v1` JWT claim", but the example immediately below has no keys under the claim, its value is a JSON list.

cdvoisin replied:
Fixed.

martin-kuba commented on researcher_ids/RI_Claims_V1.md:
```diff
@@ -60,200 +76,281 @@ The key words "MUST", "MUST NOT", "REQUIRED", "SHALL", "SHALL NOT", "SHOULD",
+    
+        1.  The Passport Clearinghouse has a sufficient trust relationship
+            with the [Passport Assertion Source](#passport-assertion-source);
+            or
```
The Clearing house should have a trust in both the Passport Assertion Source and the Passport Visa Signatory. Otherwise an evil Passport Visa Signatory may issue fake assertions pretending to be from a trusted Passport Assertion Source.

cdvoisin replied:
Done

martin-kuba reviewed (changes requested):
Just small issues, otherwise the spec seems fine to me.

cdvoisin committed:
Rename OIDC token to access token

cdvoisin committed:
Tweak definitions and trust model re Martin

Based on Martin's feedback

cdvoisin committed:
Tweak definitions and visa type descriptions

For clarity as well as removing "researcher" when a more general identity was intended.

mikael-linden commented on researcher_ids/RI_Claims_V1.md:
```diff
@@ -60,200 +76,298 @@ The key words "MUST", "MUST NOT", "REQUIRED", "SHALL", "SHALL NOT", "SHOULD",
+10. <a name="requirement-10"></a> Processing a Passport within a Passport
     Clearinghouse is to abide by the following:
+    
+    1.  Claim Clearinghouses MUST reject all requests that provide Passports
```
```suggestion
    1.  Passport Clearinghouses MUST reject all requests that provide Passports
```

cdvoisin replied:
Done

cdvoisin replied:
Done

mikael-linden commented on researcher_ids/RI_Claims_V1.md:
```diff
@@ -60,200 +76,298 @@ The key words "MUST", "MUST NOT", "REQUIRED", "SHALL", "SHALL NOT", "SHOULD",
+
+    If the Passport Broker is the [Passport Visa
+    Signatory](#passport-visa-signatory), it MUST set the "iss" to itself and
+    sign such Embedded Tokens with its own private key as described in the
```
```suggestion
    sign such Passport Visas with its own private key as described in the
```

cdvoisin replied:
Done

mikael-linden commented on researcher_ids/RI_Claims_V1.md:
```diff
@@ -60,200 +76,298 @@ The key words "MUST", "MUST NOT", "REQUIRED", "SHALL", "SHALL NOT", "SHOULD",
+            [Passport Visa Signatory](#passport-visa-signatory); or
+
+        2.  The Passport Clearinghouse can rely on a trusted service that
+            provides trust sufficient trust of any of the Passport Broker,
```
```suggestion
            provides sufficient trust of any of the Passport Broker,
```

cdvoisin replied:
Done

mikael-linden commented on researcher_ids/RI_Claims_V1.md:
```diff
@@ -279,41 +393,162 @@ interfaces, such as:
-    format of the string can vary by the
-    [claim definition](#ga4gh-researcher-identity-claim-definitions).
+-   Embedded Token Signatories, Passport Brokers, Passport Clearinghouses, and
+    their Clients MUST conform to the
```
"and their clients" Where does the scope of this spec end? I don't think we should impose requirements to actors who are not covered by this document

cdvoisin replied:
Limited the scope of client conformance to just the part that the AAI spec calls out.

mikael-linden commented on researcher_ids/RI_Claims_V1.md:
```diff
@@ -60,200 +76,298 @@ The key words "MUST", "MUST NOT", "REQUIRED", "SHALL", "SHALL NOT", "SHOULD",
-    [custom claims](#custom-researcher-identity-claim-names) and it is
-    RECOMMENDED for the page contents within the custom claim name URL to
-    define whether this field may be present or not.
+-   A GA4GH-compatible access token, as per the [GA4GH AAI
```
It would be elegant to focus on claim syntax and semantics in this spec and leave all protocol stuff ("access token") to the AAI spec.

cdvoisin replied:
As discussed, looking to address this issue in a following PR.

mikael-linden commented on researcher_ids/RI_Claims_V1.md:
```diff
@@ -60,200 +76,298 @@ The key words "MUST", "MUST NOT", "REQUIRED", "SHALL", "SHALL NOT", "SHOULD",
+    claims to only the subset needed for a particular purpose, etc.
+
+7.  <a name="requirement-7"></a> The [Passport JWT Claim](#passport-jwt-claim)
+    is only included in the Passport if the `ga4gh_passport_v1` scope is
```
It would be elegant to make this spec focus on RI claims and leave all protocol stuff ("scope") to the AAI spec.

cdvoisin replied:
Required to specify the name of the passport scope as the AAI spec leaves this as an abstraction to handle more than one spec using it.

mikael-linden commented on researcher_ids/RI_Claims_V1.md:
```diff
@@ -279,41 +393,162 @@ interfaces, such as:
+are defined in this section, other fields MAY exist within the object
+and should be ignored by any Passport Clearinghouse that is not familiar
+with the use of such fields. Field names are reserved for definition by
+the GA4GH DURI committee.
```
Can we assume DURI exists forever (or at least as long as this spec?) Would it be more future-proof to refer to GA4GH only?

cdvoisin replied:
Opened an issue to track this

mikael-linden commented on researcher_ids/RI_Claims_V1.md:
```diff
@@ -279,41 +393,162 @@ interfaces, such as:
+    [Passport Visa Expiry](#passport-visa-expiry) section.
+
+-   If a [Passport Assertion Repository](#passport-assertion-repository)
+    does not include enough information to construct an `iat` timestamp, a
```
Can you cross-check the use of "asserted" and "iat". Is the difference between the two described somewhere?

cdvoisin replied:
Fixed text in the spec. There are more details in the landing page.

cdvoisin committed:
Claim Clearinghouse to Passport Clearinghouse

cdvoisin committed:
Fix "asserted" to separate it from "iat"

cdvoisin committed:
Merge branch 'master' into master

cdvoisin committed:
Update assertion service names

And reorder Overview section.

cdvoisin committed:
Rename flow of data service names

Passport Visa Assertion Source and Passport Visa Assertion Repository

cdvoisin committed:
Added colors to flow of data diagram

cdvoisin committed:
Added legend to flow of data diagram

cdvoisin committed:
Rename Signatory and Assertion services

"Passport Visa Signatory" becomes "Passport Visa Issuer".
"Passport Assertion Repository" becomes "Passport Visa Assertion Repository".
"Passport Assertion Source" becomes "Passport Visa Assertion Source".

Some reordering and improvements to the Overview section, including an update to align with the new passport_flow_of_data_basic.svg diagram.

cdvoisin committed:
Add files via upload
